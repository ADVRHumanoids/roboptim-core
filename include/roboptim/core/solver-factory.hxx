// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Solver class.
 */

#ifndef ROBOPTIM_CORE_SOLVER_FACTORY_HXX
# define ROBOPTIM_CORE_SOLVER_FACTORY_HXX
# include <sstream>

namespace roboptim
{
  template <typename T>
  SolverFactory<T>::SolverFactory (std::string plugin, const problem_t& pb)
  throw (std::runtime_error)
    : handle_ (),
      solver_ ()
  {
    typedef solver_t* create_t (const problem_t&);

    if (lt_dlinit () > 0)
      throw std::runtime_error ("failed to initialize libltdl.");

    std::stringstream ss;
    ss << "roboptim-core-" << plugin << "-plugin";
    handle_ = lt_dlopenext (ss.str ().c_str ());
    if (!handle_)
      {
	std::stringstream sserror;
	sserror << "libltdl failed to load plug-in ``"
		<< ss.str () << "'': " << lt_dlerror ();
	throw std::runtime_error (sserror.str ().c_str ());
      }

    create_t* c =
      reinterpret_cast<create_t*> (lt_dlsym (handle_, "create"));
    if (!c)
      {
	std::stringstream sserror;
	sserror << "libltdl failed to find symbol ``create'': "
		<< lt_dlerror ();
	throw std::runtime_error (sserror.str ().c_str ());
      }
    solver_ = c (pb);

    if (!solver_)
      {
	std::stringstream sserror;
	sserror << "libltdl failed to call ``create'': "
		<< lt_dlerror ();
	throw std::runtime_error (sserror.str ().c_str ());
      }
  }

  template <typename T>
  SolverFactory<T>::~SolverFactory () throw ()
  {
    typedef void destroy_t (solver_t*);
    destroy_t* destructor =
      reinterpret_cast<destroy_t*> (lt_dlsym (handle_, "destroy"));
    if (destructor)
      {
        destructor (solver_);
        solver_ = 0;
      }

    lt_dlclose (handle_);
    lt_dlexit ();
  }

  template <typename T>
  typename SolverFactory<T>::solver_t&
  SolverFactory<T>::operator () () throw ()
  {
    assert (solver_ != 0);
    return *solver_;
  }

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_SOLVER_FACTORY_HH
