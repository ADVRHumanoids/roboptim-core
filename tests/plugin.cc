// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>

#include <boost/mpl/vector.hpp>

#include "common.hh"

#include <roboptim/core/io.hh>
#include <roboptim/core/solver-factory.hh>

using namespace roboptim;

// Specify the solver that will be used.
typedef Solver<Function, boost::mpl::vector<Function> > solver_t;

// Define a simple function.
struct F : public Function
{
  F () : Function (4, 1, "a * d * (a + b + c) + d")
  {}

  void impl_compute (result_t& result, const argument_t& argument) const throw ()
  {
    result (0) = argument[0] * argument[3]
      * (argument[0] + argument[1] + argument[2]) + argument[3];
  }

  // No gradient, hessian.
};

int run_test ()
{
  // Instantiate the function and the problem.
  F f;
  solver_t::problem_t pb (f);

  // Instantiate the factory using the dummy solver.
  SolverFactory<solver_t> factory ("dummy", pb);

  // Retrieve a solver instance from the factory.
  solver_t& solver = factory ();

  // Get the Boost.Variant minimum.
  solver_t::result_t res = solver.minimum ();

  // Get the "real" minimum.
  solver.getMinimum<SolverError> ();

  // Display problem and solver.
  std::cout << pb << std::endl
            << "---" << std::endl
            << solver << std::endl;

  // Try to get the minimum from a GenericSolver*.
  GenericSolver* gs = &solver;
  std::cout << gs->getMinimum <SolverError> ().what ()
            << std::endl;

  return 0;
}

GENERATE_TEST ()
