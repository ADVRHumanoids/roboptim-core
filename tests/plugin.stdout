Problem:
  a * d * (a + b + c) + d (not differentiable)
  Argument's bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
  Argument's scales: 1, 1, 1, 1
  No constraints.
  No starting point.
  Infinity value (for all functions): inf
---
Solver:
  Plugin: dummy
  Problem:
    a * d * (a + b + c) + d (not differentiable)
    Argument's bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
    Argument's scales: 1, 1, 1, 1
    No constraints.
    No starting point.
    Infinity value (for all functions): inf
  Result:
    Solver error: The dummy solver always fail.
  Parameters:
    dummy-parameter (dummy parameter): 42
    dummy-parameter2 (yet another dummy parameter): 3
    dummy-parameter3 (just a dummy key): ...and a dummy value!

The dummy solver always fail.
